### PR TITLE
fix: allow send_session_message and REST message delivery to reach embedded-agent workers

### PR DESCRIPTION
**Pull request by ms2sato**
## Summary

Implements PR-2 of Issue #1260 (Gap 3): `send_session_message` (MCP) and REST `POST /api/sessions/:id/messages` can now deliver to embedded-agent workers. PR-1 (auto-activation, #1262) is already merged; PR-3 (worktree query invalidation) is out of scope, tracked separately against the same Issue.

- `canReceiveSessionMessages` (`packages/shared/src/types/worker.ts`) now admits `EmbeddedAgentWorker` in addition to `AgentWorker`. `isPtyBackedWorker` is unchanged — embedded-agent workers still have no PTY; the two predicates now diverge on purpose (see glossary/design-doc updates below).
- `SessionManager.sendMessage` (same-session, backs the REST route) and the MCP `send_session_message` tool (cross-session) both gained an embedded-agent branch that converges on the same single delivery implementation: `EmbeddedAgentWorkerService.sendUserMessage`, reached via `SessionManager.sendEmbeddedAgentUserMessage`. No new service class.
- **Activate-on-delivery** (Ruling B): both branches call the existing idempotent `activateEmbeddedAgentWorker` before delivering — the same entry point PR-1's `delegate_to_worktree` path and the worker-WebSocket open handler use.
- **MCP branch delivers the same notification template** a PTY-backed target would receive (`[internal:message] ...` + `[Reply Instructions] ...`), extracted into a new pure `buildPtyNotificationText` helper shared with the (byte-identical, unchanged) PTY path — not the raw message content. The embedded agent reads the actual content from the message file with its own Read tool, exactly like a terminal agent does today.
- **REST/same-session branch delivers the raw content** (parity with the existing PTY-injection branch, which also injects raw content, not a notification wrapper), folding `filePaths` into the text since embedded delivery has no file-attachment concept.
- **TURN_IN_PROGRESS is a hard classified rejection**, not a queue (v1 has no turn queue by design) — both branches surface it via a new `EmbeddedMessageDeliveryError` marker (`packages/server/src/services/embedded-agent-worker-service.ts`), alongside the existing `EmbeddedAgentActivationError` marker for activation failures. Both REST and MCP classify: marker messages forwarded verbatim, anything else replaced by a generic fallback — mirroring PR-1's classification pattern (a 3rd local copy of `GENERIC_EMBEDDED_ACTIVATION_FAILURE_MESSAGE`, noted for future consolidation under #1259).
- Unlike the terminal/PTY branch (best-effort — the message file is already written even if the PTY notification fails), the **embedded branch is a hard failure**: the tool call / REST request fails with the classified message on activation or delivery failure. No silent drop.

## Design decisions / architect consultation

Before implementing, I consulted the architect (session 56b4d446) on Issue #1260's Ruling 1, because code investigation surfaced an ambiguity: `SessionManager.sendMessage` (same-session, backs REST) and the MCP `send_session_message` tool (cross-session, file + PTY-notification based) are two structurally different mechanisms today, and the issue's Ruling 1 text ("the embedded delivery branch belongs in `SessionManager.sendMessage`... that layer heals the REST path in the same change") read as if both should converge on literally the same method call.

The architect **corrected their own Ruling 1 phrasing**: routing the MCP branch through `SessionManager.sendMessage` (Interpretation B) would have changed terminal-agent cross-session delivery from raw `writeWorkerInput` pointer-text to `ptyMessageInjectionService.injectMessage` — a byte-identical-AC violation, and would conflate two different message semantics (pointer-file notification vs. in-session content injection). The confirmed design (Interpretation A, implemented here): the two call sites stay separate, but both converge on the same **delivery implementation** (`EmbeddedAgentWorkerService.sendUserMessage`), exactly as terminal delivery converges on its own PTY-write paths. The architect also locked three secondary specs during this exchange: (1) the MCP branch must reuse the exact same notification template as the terminal branch, not inline full content; (2) activate-on-delivery applies uniformly to both branches, not just MCP; (3) TURN_IN_PROGRESS rejects with the existing classification — no queueing.

**Q11 (Ruling C) — tool-surface four-question audit.** `send_session_message` is a session-messaging operation, not an agent-directory operation, so `AGENT_OPERATIONS` (`packages/shared/src/types/agent-operations.ts`) and its 3 exposure tables are **not touched** — force-fitting an entry there was explicitly rejected by the architect. Applying Q11 as the general tool-surface audit instead: **this PR closes the asymmetry where the embedded-agent surface could originate a delegation callback (via `delegate_to_worktree`'s auto-appended callback instructions) but could not receive one** — the delegated embedded agent's `send_session_message` call back to an embedded-agent parent used to reject with `cannot receive inbound messages: requires a PTY-backed worker (agent/terminal)`; it now succeeds.

**Predicate audit (`canReceiveSessionMessages` / `isPtyBackedWorker` call sites).** `canReceiveSessionMessages` had exactly one pre-existing call site (`mcp-server.ts`'s auto-select filter) — confirmed still the only one needing the widened behavior. `isPtyBackedWorker` has 4 call sites in `mcp-server.ts`; only the `send_session_message` explicit-`toWorkerId` guard changed (now `!isPtyBackedWorker(worker) && !canReceiveSessionMessages(worker)`, admitting embedded-agent without touching `isPtyBackedWorker`'s own definition); the other 3 (`create_timer`, `create_conditional_wakeup`, `run_process`) gate genuinely PTY-only tools and are untouched. `canReceiveSessionMessages` intentionally does **not** add `'terminal'` — the auto-select filter's pre-existing exclusion of terminal workers is a separate, unrelated asymmetry, out of scope here.

**Q8 (signature-change call-site estimate).** `SessionManager.sendMessage` becomes `async`. Call-site count: 1 production (`routes/workers.ts:268`) + 12 test call sites in `session-manager.test.ts` (all `await`-ed). `InterSessionMessageService.sendMessage` (a different method, used by `process-output-router.ts`) is unrelated and untouched.

**Design-doc / glossary sync** (`docs/design/embedded-agent-worker.md`, `docs/glossary.md`): the design doc had explicitly pre-declared this exact widening as post-v1 fast-follow item #3 (`extend canReceiveSessionMessages`) — marked done in this PR, with its stale code snippet refreshed. The glossary's "PTY-backed Worker" entry stated `canReceiveSessionMessages` as an alias-equivalent of the PTY-backed grouping; updated to describe the now-intentional divergence.

## Verification

- `bun run typecheck && bun run test` (full suite, independently re-run after reviewing the diff, not just trusting the implementing agent's report): server 3752 pass / 1 skip (environment-gated, pre-existing) / 0 fail, client 2107 pass, shared 597 pass, integration 73 pass, embedded-agent 295 pass, scripts 521 pass. `TEST_EXIT: 0`.
- **TDD polarity** (per Issue AC): the implementing agent stashed only the production-file hunks (tests kept) and ran the new embedded-branch `mcp-server.test.ts` cases — all failed, the first with the exact pre-fix text `cannot receive inbound messages: requires a PTY-backed worker (agent/terminal)`. Restored the fix; same tests passed (positive control). I independently re-read the diff against this description and confirmed the `else` branch in `mcp-server.ts`'s `send_session_message` handler is byte-identical to the pre-fix code (only wrapped in an `if/else`).
- `node .claude/skills/orchestrator/preflight-check.js`: exit 0. Coverage: 3/3 tracked files covered. One advisory (non-blocking) note: "Integration test gap detected" for `routes/workers.ts` / `shared/types/worker.ts`. This PR does not add a `packages/integration/` test; the wire boundary is instead exercised by real-protocol tests (`app.request` through the actual Hono MCP app in `mcp-server.test.ts`, and through the actual Hono REST route in `routes/__tests__/workers.test.ts`) plus the live E2E below against a real running server. No new wire schema field was added (Q10 N/A — `canReceiveSessionMessages` is a predicate function, not a serialized field), so the schema-drift risk Q10 guards against doesn't apply here. Flagging this explicitly per `workflow.md`'s joint-decision requirement for the Orchestrator/architect to confirm before merge.
- CodeRabbit CLI: installed but `signed out` in this headless worktree (same limitation as PR #1262) — not a rate-limit, will not self-resolve by waiting.

## Live E2E on isolated instance

Per the established pattern (PR #1262's "Live E2E on isolated instance"): an isolated single-user instance was started from this worktree, custom ports, fresh `AGENT_CONSOLE_HOME`, throwaway git repo, and a small inert-but-responsive stub OpenAI-compatible provider — the shared dev server (3457/5173/6340) was never touched.

```
mkdir -p ~/.agent-console-dev-1260pr2
PORT=17280 CLIENT_PORT=17281 AGENT_CONSOLE_HOME=~/.agent-console-dev-1260pr2 bun run dev
# stub provider on :17282
```

**Cross-session MCP path (the Issue's core AC item 4 — child delegate reporting back to an embedded-agent parent):** created a parent `quick` session with a deactivated embedded-agent worker, and a separate child session. Called `send_session_message` via raw MCP JSON-RPC (`POST /mcp`, `tools/call`) exactly as a real MCP client would, targeting the parent's embedded worker:

- Tool call succeeded (no `isError`).
- `GET /api/sessions/<parent>` showed the embedded worker's `activated` flip from `false` to `true` — activate-on-delivery fired live.
- `ps` confirmed a real OS subprocess spawned (`bun .../packages/embedded-agent/src/main.ts`).
- The stub provider's request log shows the parent's actual conversation turn sent to the LLM containing the exact notification template:
  ```
  [internal:message] timestamp=... source=session from=<child-session-id> summary="Message from session <child-session-id>" path=<message-file-path> intent=triage
  [Reply Instructions] To reply, use the send_session_message MCP tool with:
  - toSessionId: "<child-session-id>"
  - fromSessionId: Use your AGENT_CONSOLE_SESSION_ID environment variable
  ```
  — this is the message reaching the parent's actual conversation, verifiable in the same way a terminal-agent's PTY notification is (in this case, as the literal turn text sent to the model).

**Same-session REST path (heals the REST endpoint in the same change, per Ruling 1):** created a second session with a deactivated embedded-agent worker, `POST /api/sessions/:id/messages` (multipart form, mirroring the browser Message Panel) targeting it directly with raw content (no notification wrapper, matching PTY-branch parity) — `201`, `activated` flipped to `true`, and the stub provider's log shows the raw content (`"Hello via REST composer to an embedded worker..."`) delivered as-is, confirming the REST/UI same-session path is no longer PTY-only.

**Terminal-agent regression:** not re-driven live in this E2E (the `else` branch in both `session-manager.ts` and `mcp-server.ts` is byte-identical to pre-fix code, confirmed by diff review above) — covered by the full, unmodified pre-existing terminal-agent `send_session_message` / REST-message test suites, which all pass unchanged.

**Teardown:** killed the isolated instance's full process tree (server, vite, both embedded-agent subprocesses, stub provider) and verified via `ps`/`ss -tln` that ports 17280/17281/17282 are free. `~/.agent-console-dev-1260pr2` data dir left in place (throwaway, not shared state).

## Scope guards honored

`AGENT_OPERATIONS` untouched; generic-activation-failure-message literal consolidation left for #1259; `isPtyBackedWorker`'s definition and its other 3 call sites untouched; `canReceiveSessionMessages` does not admit `'terminal'`; no message queue for busy embedded targets (`TURN_IN_PROGRESS` is a hard reject).

## Test plan

- [x] `bun run typecheck && bun run test` — full suite green, exit 0 (see Verification above).
- [x] `node .claude/skills/orchestrator/preflight-check.js` — exit 0, advisory integration-test-gap note flagged above for Orchestrator/architect review.
- [x] TDD polarity confirmed for the new embedded-branch tests.
- [x] Live E2E on an isolated instance — both the MCP cross-session path and the REST same-session path verified against a real running server with a real spawned subprocess.
- [x] CodeRabbit review: genuine walkthrough obtained, 2 findings raised and fixed, both marker-closed as addressed (see "CodeRabbit disposition" below).

## CodeRabbit disposition

The GitHub-side bot posted a genuine walkthrough on the initial commit (2026-07-31T21:05:25Z, `COMMENTED`) with 2 findings, both verified as legitimate independently and fixed in commit `9ffea454`:

- **Major** (`packages/server/src/routes/workers.ts`): the catch block around `sessionManager.sendMessage` classified every thrown error as an embedded-agent activation failure, downgrading unrelated errors (PTY-branch / shared-path failures) from a proper 5xx into a misleading 400. Fixed: marker errors (`EmbeddedAgentActivationError` / `EmbeddedMessageDeliveryError`) forwarded verbatim; non-marker error classified with the generic fallback only when the target worker actually resolves to `embedded-agent`; everything else rethrown unchanged. Regression test added (TDD-polarity-verified: 400 pre-fix, 500 post-fix for an unrelated error against a PTY-backed target).
- **Minor**: a `pty-notification.test.ts` case asserted against an array that could never be populated (vacuous assertion). Fixed: spies on `setTimeout` directly and asserts it's never called.

Both findings' reply threads carry the bot's own `<review_comment_addressed>` marker after independently re-reading the fix commit's actual file contents (not just accepting the reply text) — see PR discussion for the verification detail. Current state at commit `9ffea454`: `reviewDecision` empty (expected — the bot only ever submits `COMMENTED`-state reviews in this thread, never a formal top-level `APPROVED`/`CHANGES_REQUESTED` event), 0 open actionable inline comments, and the commit-status `description` reads `Review rate limited` (the bot's formal full-walkthrough re-pass is rate-limited for this exact head SHA, distinct from its incremental reply-analysis capability, which remained functional and did perform real analysis against this commit). Per `coderabbit-ops` skill's "walkthrough-exists rubric" (a genuine walkthrough exists + inline actionable comments = 0 + `reviewDecision` empty → treat as clean-equivalent): this PR meets that bar. Local CLI remains a headless-worktree auth-fail (structural, unrelated to rate-limiting) throughout, as noted in the original CodeRabbit CLI note above.

## Preflight advisory disposition (joint decision, architect-ruled)

The preflight "Integration test gap detected" advisory (flagged in Verification above) was reviewed by the architect (session 56b4d446) and ruled resolved, not skipped:

`canReceiveSessionMessages` is not a derived field crossing the wire — it is a compile-time predicate (a function), never serialized. It is therefore structurally not exposed to the #926-category hazard (`workflow.md` Q10: valibot's default permissive parse silently stripping an unknown derived field). The actual wire-level contract this PR changes — the MCP tool's target-worker acceptance behavior, and the REST route's classified error responses — is exercised end-to-end by real-protocol tests: `mcp-server.test.ts`'s new embedded-branch cases drive the actual `createMcpApp` Hono app via `tools/call`, and `routes/__tests__/workers.test.ts`'s new cases drive the actual REST route via `app.request`, plus both paths are additionally verified live against a real running server in "Live E2E on isolated instance" above.

Per `workflow.md`'s "the joint decision can move *which* test satisfies a `strongly recommended` warning, not *whether* one is needed" — `whether` is satisfied; `which` moved from a `packages/integration/src/` test to the MCP protocol tests + REST route tests + live E2E, because the risk category the advisory exists to catch does not apply to this change's shape (a predicate widening, not a wire schema addition).

Refs #1260

## Summary by CodeRabbit

* **New Features**
  * Session messages can now be delivered to embedded-agent workers.
  * Deactivated embedded agents are activated automatically when a message is sent.
  * File paths included in messages are converted into readable text.
* **Bug Fixes**
  * Added clearer handling for delivery failures, including busy, inactive, and write errors.
  * Unexpected activation failures now return a safe, generic error message.
* **Tests**
  * Expanded coverage for embedded-agent messaging, notification formatting, activation, and error handling.

### CODE REVIEW

File: packages/server/src/routes/workers.ts [hunk 1]
```diff
@@ -265,7 +285,19 @@ const workers = new Hono<AppBindings>()
...
+    } catch (err) {
+      // Clean up saved files since the message was not delivered
+      await Promise.allSettled(savedPaths.map((p) => unlink(p)));
+      const classified =
+        err instanceof EmbeddedAgentActivationError || err instanceof EmbeddedMessageDeliveryError
+          ? err.message
+          : GENERIC_EMBEDDED_ACTIVATION_FAILURE_MESSAGE;
+      logger.warn({ err, sessionId, toWorkerId: validated.toWorkerId }, 'Embedded-agent message delivery failed');
+      throw new ValidationError(classified);
+    }
```
[the remark this answers is not in the document]
Comment (ms2sato) [reply packages/server/src/routes/workers.ts:321]: Fixed — the catch block now only classifies `EmbeddedAgentActivationError`/`EmbeddedMessageDeliveryError` marker types verbatim. For non-marker errors, the target worker's type is looked up via `sessionManager.getWorker`: if it's an embedded-agent worker, the (non-marker) error is still classified with the generic fallback message (activation can genuinely throw a bare `Error` from e.g. provider-key loading or spawn — see `EmbeddedAgentActivationError`'s JSDoc), preserving the existing 'classified 400 for non-marker embedded reason' test. Otherwise (a genuinely unrelated/shared-path failure) the error is rethrown unchanged so `onApiError` reports the real status/message instead of a misleading 400. Added a regression test (TDD-verified: fails with 400 pre-fix, passes with 500 post-fix) mocking `sessionManager.sendMessage` to reject with an unrelated error for a PTY-backed target worker. Confirmed `mcp-server.ts`'s `send_session_message` embedded branch does not have this bug — its try/catch is already narrowly scoped to only the `activateEmbeddedAgentWorker` call, gated on `resolvedWorker?.type === 'embedded-agent'` beforehand.

File: packages/server/src/lib/__tests__/pty-notification.test.ts [hunk 2]
```diff
@@ -233,3 +233,52 @@ describe('writePtyNotification', () => {
...
+        tag: 'internal:message',
+        fields: { source: 'session', from: 'sender-1', summary: 'hi', path: '/tmp/msg' },
+        intent: 'inform',
+      });
+      jest.advanceTimersByTime(1000);
+      expect(written).toHaveLength(0);
+    } finally {
+      jest.useRealTimers();
+    }
+  });
```
[the remark this answers is not in the document]
Comment (ms2sato) [reply packages/server/src/lib/__tests__/pty-notification.test.ts]: Fixed — replaced the fake-timer scaffolding with a direct spyOn(globalThis, 'setTimeout') assertion. The test now asserts setTimeout was never called, instead of the previous vacuous check on an untouched local written array.

File: packages/server/src/routes/workers.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ms2sato) [reply packages/server/src/routes/workers.ts:321]: The fix commit (9ffea454) had not been pushed yet when this analysis ran — it was only committed locally at that point. Pushed now. The route's catch block (packages/server/src/routes/workers.ts:288-321) now does: marker types forwarded verbatim; non-marker error + sessionManager.getWorker(sessionId, validated.toWorkerId)?.type === 'embedded-agent' -> classified with the generic fallback (a bare Error can legitimately come from EmbeddedAgentWorkerService.runActivation, not just the marker class); otherwise rethrown unchanged so onApiError reports the real status/message. Regression test added in workers.test.ts covering the unrelated-error-for-a-PTY-backed-target case. Please re-check against the current HEAD (9ffea454).
